export_snowflake/__init__.py

```diff
@@ -521,6 +519,134 @@ def flush_records(stream: str,
+    # https://boto3.amazonaws.com/v1/documentation/api/latest/guide/clients.html#multithreading-or-multiprocessing-with-clients
+    session = boto3.session.Session()
+    s3_client = session.client('s3')
+    
+    with concurrent.futures.ThreadPoolExecutor(max_workers=6) as executor:
+        futures = [executor.submit(upload_to_snowflake_per_thread, s3_client, csv_list, error_queue, config, schema, file_format_type) for csv_list in s3_csv_lists]
+        
+        for future in concurrent.futures.as_completed(futures):
+            try:
+                future.result()
+            except Exception as exc:
+                raise SymonException(f'An error occurred while loading data to Snowflake: ' + exc, 'snowflake.clientError')
+
+    parallel_transfer_end_time = time.time()
+    stream = config["stream"]
+    LOGGER.info(f"Elapsed time usage for {stream} is {parallel_transfer_end_time - parallel_transfer_start_time}")
+
+
+def export_to_snowflake(config, o, file_format_type): 
+    # retrieve the csv.gz lists and the schema file from the s3 bucket
+    s3_client = boto3.client('s3')
+
+    bucket = config["bucket"]
+    prefix = config["prefix"]
+    response = s3_client.list_objects_v2(Bucket=bucket, Prefix=prefix)
+
+    s3_csv_lists = []
+    schema_file = None
+    for obj in response['Contents']:
+        if obj['Key'].endswith('.csv.gz'):
+            # append the exportTaskID to the file name to avoid conflict
+            file_name = obj['Key'].split("/")[-2] + '-' + obj['Key'].split("/")[-1]
+            s3_csv_lists.append((bucket, obj['Key'], file_name, config["stream"]))
+        # for both python and spark parquet_to_csv steps, we always have a schema.json with the schema details
+        if obj['Key'].endswith('.json'):
+            schema_file = obj['Key']
+
+    if schema_file is None:
+        raise Exception("Schema file not found in the s3 bucket")
+    if len(s3_csv_lists) == 0:
+        raise Exception("No csv files found in the s3 bucket")
+
+    local_schema_file = 'local_schema.json'
+    s3_client.download_file(bucket, schema_file, local_schema_file)
+
+    f = open(local_schema_file)
+    schema = json.load(f)
+
+    # generate the schema object for DbSync
+    stream = config["stream"]
+    o = {
+        "stream": stream,
+        "schema": {"properties": schema['properties']},
+        "key_properties": config["key_columns"]
+    }
+
+    # check if the schema(Snowflake table) exists and create it if not
+    # sync_table will check the schema with the table in Snowflake, if there's a miss matching in columns, raise an error
+    db_sync = DbSync(config, o, None, file_format_type)
+    db_sync.create_schema_if_not_exists()
+    db_sync.sync_table()
+
+    # the csv files in s3 are compressed, changed the parquet_to_csv to write into gzip format, also remove the headers
+    transfer_data_from_s3_to_snowflake(s3_csv_lists, config, o, file_format_type)
```

Comment:
When we use stage, there's storage costs associated with it.
https://docs.snowflake.com/en/user-guide/cost-understanding-data-storage

This article summarizes how the storage usage is calculated.
https://community.snowflake.com/s/article/How-Snowflake-Charges-for-Data-Storage

In the article, it says it takes hourly snapshot of all customer data and then average this across each day when they calculate the daily average usage in bytes. I think it might be worth to check how hourly snapshot is affected if we
- upload one file > load to table > delete from stage on each iteration
- upload all files > load to table on each iteration > delete all files from stage
I think with second approach, hourly snapshot might be higher.

Reply:
no actual stage files applied after the change to transfer data from s3 to snowflake directly.